Oracle.md

```diff
@@ -0,0 +1,153 @@
+This necessary information is committed in a so-called _event descriptor_ that will be further detailed in this document.
+Descriptors are classified and discovered through a namespace presently represented as Unique Resource Identifier (URI) paths.
+
+## Event descriptor
+
+An event descriptor provides information to clients about an event for which an oracle plans on releasing a signature over its outcome.
+The provided information should be sufficient for a client to create a set of CETs that will cover all possible outcomes of the event.
+
+Regardless of the outcome type, an event descriptor should include:
+
+- The R value that will be used to sign the event outcome (possibly several).
+- The public key of the oracle.
+- The earliest date at which the event outcome will be released in Unix epoch format.
+
+Here we assume that an event outcome can be represented either as a set of strings or numbers.
+Three kinds of event outcomes are defined (note that a single even can be composed of several types).
+
+### Simple enumeration
+
+For events that have a narrow range of possible outcomes, the outcomes can simply be enumerated.
+For the sake of simplicity, outcomes should be represented as strings.
+
+#### Example: Weather tomorrow
+
+Tomorrow's weather can be represented as the set of strings `[sunny, cloudy, rainy]`.
+
+### Range
+
+When an event has numerical outcomes that cannot be easily enumerated, they can be represented as a series with:
+
+- start: the first possible outcome number
+- end: the last possible outcome number
+- step: the increment
+
+#### Example: tomorrow's temperature
+
+```
+start: -100
+end: 100
+step: 1
+```
+
+### Large range
+
+When a range of a numerical outcomes is large or unbounded, the oracle can represent the outcome using numerical decomposition.
+
+The event descriptor should include:
+
+- base: the base in which the outcome value is decomposed.
+
+Prior to the event, the oracle releases `n + 1` R-values, where `n` is the expected number of digits that will make up the outcome value.
+The oracle releases an additional R-value with which it will sign the actual number of digits that make up the outcome value.
+This last value is meant for covering cases where the outcome value goes outside of the range that it was expected to stay in.
+The oracle should provide the R values as a list where the extra R value is placed first.
+
+#### Example: BTC/USD rate
+
+Alice and Bob wish to enter into a DLC at time `t` when BTC/USD ~= $10000 with a maturity time of `t+2`.
+The oracle at that time has a safety margin of one digit, and so provides 1 + 6 R-values:
+
+```
+base: 10
+RValues: [R0, R1, R2, R3, R4, R5]
+```
+
+Alice and Bob create a DLC with "relevant" outcomes between $10000 and $11000.
+They use the R0 value to cover the cases where BTC/USD < $10000 and BTC/USD > $99999.
+As they want to be safe, they cover cases up to a value of BTC/USD composed of 10 digits, which they can do with only a few extra CETs.
```

Comment:
It might be good to explicitly enumerate the CETs for this case.  I think I see how to efficiently construct `log(n)` of them, given that each CET can represent a boolean `AND` of several R values.  But I'm not sure that it will be obvious to many readers.

I assume the basic idea is to construct CETs as follows:

- `R0=1` - outcome is underflow
- `R0=2` - outcome is underflow
...
- `R0=5 && R1=1 && R2=0` - this is the desired range, but this needs to be further decomposed, assuming the distribution to each party is different based on the exact price event outcome
- `R0=5 && R1=1 && R2>0` - 9 overflow cases
- `R0=5 && R1>1` - 8 more overflow cases
- `R0=6` - overflow
...
- `R0=10` - overflow

Reply:
It might be good to represent both equality and open-ended range for the exponent, which allows for complete safety.  i.e.  could represent `R0=5` and also `R0_ge=5` (representing the fact that the exponent is greater or equal to 5 using a different R value).  This allows a CET that deals with arbitrary overflows.

Reply:
Thanks for the suggestion, I updated roughly following it.

> It might be good to represent both equality and open-ended range for the exponent, which allows for complete safety. i.e. could represent R0=5 and also R0_ge=5 (representing the fact that the exponent is greater or equal to 5 using a different R value). This allows a CET that deals with arbitrary overflows.

That's a good suggestion, though one issue I can see with this is that it makes it difficult for an oracle to update an event as I describe in the example. So if at time `t` the oracle releases 5 R values, but at time `t+1` because of large price increase he wants to now use 6 for newly created contract, the `R0_ge=5` is not useful for these newly created contracts. Maybe it could be possible to at that time create yet another `R0_ge` to be used for when price goes above 6 digits, but from an implementation perspective it feels like becoming a bit complex, though you are right that this would provide complete safety which is nice.

It would of course also be possible for the oracle to just create a different event, but it also makes things more complicated IMHO.

Reply:
Ah, right, I should have called it `R_ge_5`, since the oracle can't safely sign more than one thing with a single R value.

Also, it may be that exponential increases in an event value are so unlikely that this is overkill.

Reply:
It might be an overkill, but at the same time it's also a bit neater. I haven't fully made up my mind but I think it's really worth considering.

Reply:
Another potential approach is related to your idea about different events.  Any particular event could have a "saturation point" - e.g. if the Oracle doesn't expect the price to go above 99999, then part of the event definition could be that it saturates at that point.  i.e. the oracle signs the limit if the real-world event goes above it.

If things move strongly in that direction, then a new event could be created with a higher saturation point.  But that doesn't mean the existing contracts missed out on anything, because they already had saturated outcomes in their CETs, most likely at an even lower value.

Reply:
Yeah I'm starting to think this might be the direction to go, though instead of saturation point I would maybe still keep an extra R0 that can be just 0 or 1 (1 meaning it's overflowed). In that case I think we need to think about having possible versions for events, so that we can have a same "base ID" (which for example would represent BTC/USD at a certain date for a given exchange) but multiple versions in case of large price fluctuations. One thing we lose is the small optimization in the number of CETs but as @LLFourn [noted](https://github.com/discreetlogcontracts/dlcspecs/issues/65#issuecomment-692483053) it's quite small.

I will try to update tomorrow with this approach, let me know if you have other comments.

Reply:
Finally got around to update this, @devrandom if you have time to have a look, I finally used the "sign the highest possible number" approach  as you suggested (I think it's the easiest and also cleanest).

Reply:
:+1: